### PR TITLE
🐛 Fix user defined test bugs

### LINKED ISSUE #277: Fix typo in user defined test fail msg
This `❌  User defined data validation tests passed!`  should be `❌  User defined data validation tests failed!`

### PR DESCRIPTION
**Pull request by znatty22**
Closes #277 and #278

### CODE REVIEW

File: kf_lib_data_ingest/etl/ingest_pipeline.py [hunk 1]
```diff
@@ -263,11 +263,22 @@ def user_defined_tests(self):
...
+        # Check if exit code is one of the codes that should result in pass
+        # Exit code 0:	All tests were collected and passed successfully
+        # Exit code 5:	No tests were collected
+        passed = exit_code in {0, 5}
+
+        if exit_code == 0:
+            self.logger.info(f'✅  User defined data validation tests passed')
+        elif exit_code == 5:
+            self.logger.warning(
+                f'⚠️  pytest did not collect any user defined tests, '
```
Comment (znatty22) [inline kf_lib_data_ingest/etl/ingest_pipeline.py]: Lol, its weird it doesn't properly render the unicode. Should be ⚠️

Comment (fiendish) [reply kf_lib_data_ingest/etl/ingest_pipeline.py]: ??? 
Screen Shot 2019-04-29 at 6 50 38 PM

Comment (znatty22) [reply kf_lib_data_ingest/etl/ingest_pipeline.py]: whaaaat weird! 
image

File: kf_lib_data_ingest/etl/ingest_pipeline.py [hunk 2]
```diff
@@ -263,11 +263,22 @@ def user_defined_tests(self):
...
-            self.logger.info(f'✅  User defined data validation tests passed!')
+        exit_code = pytest.main([user_defined_test_dir])
+
+        # Check if exit code is one of the codes that should result in pass
+        # Exit code 0:	All tests were collected and passed successfully
+        # Exit code 5:	No tests were collected
+        passed = exit_code in {0, 5}
+
+        if exit_code == 0:
+            self.logger.info(f'✅  User defined data validation tests passed')
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/ingest_pipeline.py]:
```suggestion
            self.logger.info(f'✅ User defined data validation tests passed')
```

lol, fix your terminal

File: kf_lib_data_ingest/etl/ingest_pipeline.py (hunk 1 above)
Comment (fiendish) [inline kf_lib_data_ingest/etl/ingest_pipeline.py]:
```suggestion
                f'⚠️ pytest did not collect any user defined tests, '
```

File: kf_lib_data_ingest/etl/ingest_pipeline.py [hunk 3]
```diff
@@ -263,11 +263,22 @@ def user_defined_tests(self):
...
+        elif exit_code == 5:
+            self.logger.warning(
+                f'⚠️  pytest did not collect any user defined tests, '
+                'even though user tests directory exists: '
+                f'{user_defined_test_dir}')
         else:
-            self.logger.info(f'❌  User defined data validation tests passed!')
+            self.logger.info(f'❌  User defined data validation tests failed '
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/ingest_pipeline.py]:
```suggestion
            self.logger.info(f'❌ User defined data validation tests failed '
```

File: tests/test_user_def_tests.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        content = '''def test():
+            assert True'''
+        test_file.write(content)
+    assert os.path.isfile(filepath)
+
+    ingest_pipeline = make_ingest_pipeline(
+        config_filepath=ingest_pkg_dir,
+        log_dir=os.path.join(ingest_pkg_dir, 'logs'))
+    assert ingest_pipeline.run()
+    assert '✅  User defined data validation tests passed' in info_caplog.text
```
Comment (fiendish) [inline tests/test_user_def_tests.py]:
```suggestion
    assert '✅ User defined data validation tests passed' in info_caplog.text
```

File: tests/test_user_def_tests.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+def test_user_def_test_not_collected(info_caplog, ingest_pkg_dir):
+    """
+    Test successful pass of user defined tests even when no tests collected
+    """
+    # Run pipeline and check log
+    ingest_pipeline = make_ingest_pipeline(
+        config_filepath=ingest_pkg_dir,
+        log_dir=os.path.join(ingest_pkg_dir, 'logs'))
+    assert ingest_pipeline.run()
+    assert (f'⚠️  pytest did not collect any user defined tests' in
```
Comment (fiendish) [inline tests/test_user_def_tests.py]:
```suggestion
    assert (f'⚠️ pytest did not collect any user defined tests' in
```

File: tests/test_user_def_tests.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            assert False'''
+        test_file.write(content)
+    assert os.path.isfile(filepath)
+
+    # Run pipeline and check log
+    ingest_pipeline = make_ingest_pipeline(
+        config_filepath=ingest_pkg_dir,
+        log_dir=os.path.join(ingest_pkg_dir, 'logs'))
+    assert not ingest_pipeline.run()
+    assert (f'❌  User defined data validation tests failed' in
```
Comment (fiendish) [inline tests/test_user_def_tests.py]:
```suggestion
    assert (f'❌ User defined data validation tests failed' in
```

File: tests/test_user_def_tests.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def test_user_def_test_fail(info_caplog, ingest_pkg_dir):
+    """
+    Test failure of user defined tests
+    """
+    # Create failing test file
+    filepath = os.path.join(ingest_pkg_dir, 'tests', 'test_passing.py')
+    with open(filepath, 'w') as test_file:
+        content = '''def test():
```
Comment (fiendish) [inline tests/test_user_def_tests.py]: I think you should use \n instead of '''...''' with a line split

File: tests/test_user_def_tests.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def test_user_def_test_pass(info_caplog, ingest_pkg_dir):
+    """
+    Test successful user defined tests
+    """
+    # Create passing test file
+    filepath = os.path.join(ingest_pkg_dir, 'tests', 'test_passing.py')
+    with open(filepath, 'w') as test_file:
+        content = '''def test():
```
Comment (fiendish) [inline tests/test_user_def_tests.py]: I think you should use \n instead of '''...''' with a line split

**Comment by fiendish** [comment]
see #282

**Comment by znatty22** [comment]
Incorporated #282

**Pull request merged**